field/TinyLang/Field/Raw/Core.hs

```diff
@@ -0,0 +1,71 @@
+{-| A Raw (untyped) AST
+-}
+module TinyLang.Field.Raw.Core
+    ( Identifier
+    , Var(..)
+    , Expr(..)
+    , BinOp(..)
+    , UnOp(..)
+    , Statement(..)
+    , RawExpr
+    , RawStatement
+    ) where
+
+import TinyLang.Field.UniVal
+import Data.Field
+
+{-| = AST
+-}
+type Identifier = String
+
+newtype Var = Var Identifier
+    deriving (Eq, Show)
+
+{-| @Expr v f@ is parameterised by the type of variable @v@.
+-}
+data Expr v f
+    = EConst     (SomeUniVal f)
+    | EVar       v
+    | EAppBinOp  BinOp           (Expr v f) (Expr v f)
+    | EAppUnOp   UnOp            (Expr v f)
+    | EStatement (Statement v f) (Expr v f)
+    | EIf        (Expr v f)      (Expr v f) (Expr v f)
+
+data BinOp
+    = Or
+    | And
+    | Xor
+    | FEq
+    | FLe
+    | FLt
+    | FGe
+    | FGt
+    | Add
+    | Sub
+    | Mul
+    | Div
+    | BAt
+
+data UnOp
+    = Not
+    | Neq0
+    | Neg
+    | Inv
+    | Unp
+
```

Comment:
I'd write `deriving (Show)` here instead of having a standalone deriving declaration at the bottom as the latter results in a lot of boilerplate when there are many instances to derive.

Applies to other data types as well.